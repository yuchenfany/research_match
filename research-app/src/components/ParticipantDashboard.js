/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import React, { useState, useEffect } from 'react';
import '../assets/index.css';
import NavBar from './NavBar';

import { getUserTags, getUserInfo } from '../modules/user-api';
import { getStudyByTag } from '../modules/study-api';
import { getNumMessagesSent } from '../modules/chat-api';

function ParticipantDashboard({ user }) {
  const [totalCompensation, setTotalCompensation] = useState(0);
  const [numEnrolled, setNumEnrolled] = useState(0);
  const [numRecommended, setNumRecommended] = useState(0);
  const [numMessages, setNumMessages] = useState(0);

  // get all studies
  async function getNumRecommended() {
    const tags = await getUserTags(user);
    const studiesByTag = await Promise.all(tags.map((tag) => getStudyByTag(tag)));
    const allStudies = studiesByTag.flat();
    return allStudies?.length ?? 0;
  }

  async function getNumStudies() {
    const data = await getUserInfo(user);
    return data?.enrolled?.length ?? 0;
  }

  async function getTotalCompensation() {
    const data = await getUserInfo(user);
    return data?.compensation?.length ?? 0;
  }

  useEffect(() => {
    getNumStudies()
      .then(setNumEnrolled);
  }, []);

  useEffect(() => {
    getTotalCompensation()
      .then(setTotalCompensation);
  }, []);

  useEffect(() => {
    getNumRecommended()
      .then(setNumRecommended);
  }, []);

  useEffect(() => {
    getNumMessagesSent(user)
      .then(setNumMessages);
  }, []);

  const cards = [
    ['TOTAL COMPENSATION EARNED', `$${totalCompensation}`],
    ['TOTAL ENROLLED STUDIES', `${numEnrolled} ${numEnrolled === 1 ? 'study' : 'studies'}`],
    ['TOTAL RECOMMENDED STUDIES', `${numRecommended} ${numRecommended === 1 ? 'study' : 'studies'}`],
    ['TOTAL MESSAGES SENT', `${numMessages} ${numMessages === 1 ? 'message' : 'messages'}`],
  ].map(([name, content]) => (
    <div key={name} className="dash-card">
      <h4>{name}</h4>
      <p>{content}</p>
    </div>
  ));

  if (!user.username) {
    return <div>forbidden</div>;
  }

  return (
    <div className="Dashboard">
      <NavBar user={user} />
      <div className="dashboard">
        <h1 className="header-left">Analytics</h1>
        <div className="dash-flex">{cards}</div>
      </div>
    </div>
  );
}

export default ParticipantDashboard;
