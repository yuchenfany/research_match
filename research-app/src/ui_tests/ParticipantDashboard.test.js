/**
 * @jest-environment jsdom
 */

/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
// import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import React from 'react';
// import { jest } from '@jest/globals';

import ParticipantDashboard from '../components/ParticipantDashboard';

// const mockedUsedNavigate = jest.fn();

// jest.mock('react-router-dom', () => ({
//   ...jest.requireActual('react-router-dom'),
//   useNavigate: () => mockedUsedNavigate,
// }));

test('Title Analytics', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ParticipantDashboard user={user1} />
        </Router>);  
  const label = screen.getByText("Analytics");
  expect(label).toBeInTheDocument();
});

test('TOTAL COMPENSATION EARNED text', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ParticipantDashboard user={user1} />
        </Router>);  
  const label = screen.getByText("TOTAL COMPENSATION EARNED");
  expect(label).toBeInTheDocument();
});

test('TOTAL ENROLLED STUDIES text', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ParticipantDashboard user={user1} />
        </Router>);  
  const label = screen.getByText("TOTAL ENROLLED STUDIES");
  expect(label).toBeInTheDocument();
});

test('TOTAL RECOMMENDED STUDIES', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ParticipantDashboard user={user1} />
        </Router>);  
  const label = screen.getByText("TOTAL RECOMMENDED STUDIES");
  expect(label).toBeInTheDocument();
});

test('TOTAL MESSAGES SENT', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ParticipantDashboard user={user1} />
        </Router>);  
  const label = screen.getByText("TOTAL MESSAGES SENT");
  expect(label).toBeInTheDocument();
});
// test('Update Button',() => {
//   const Button = ({onClick, children}) => (
//     <button onClick={onClick}>{children}</button>
//   )
//   const mockSetUser = jest.fn();
//   const user1 = { username: 'user1', password: 'user1pass' };
//     render(
//         <Router>
//             <ParticipantDashboard user={user1} />
//         </Router>);  
//   const linkElement = screen.getByRole('button', {
//     name: /update/i,
//   });
//   const handleClick = jest.fn()
//   render(<Button onClick={handleClick}>updatee</Button>)
//   fireEvent.click(screen.getByText(/updatee/i));
//   expect(linkElement).toBeInTheDocument();
// });


test('snapshot test', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
  const component = renderer.create( 
    <Router>
        <ParticipantDashboard user={user1} />
    </Router>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
//TO GET HIGHER SCORE

//1. Get the json values out from the dashboard and test the values are there
