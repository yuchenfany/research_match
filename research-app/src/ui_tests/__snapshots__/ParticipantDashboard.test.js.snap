// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="Dashboard"
>
  <div
    className="nav"
  >
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      HOME
    </button>
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      MY STUDIES
    </button>
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      MESSAGES
    </button>
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      ANALYTICS
    </button>
    <button
      className="account-btn"
      onClick={[Function]}
      type="button"
    >
      ACCOUNT
    </button>
    <div
      className="acct-options"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <button
        className="account-btn"
        onClick={[Function]}
        type="button"
      >
        EDIT PROFILE
      </button>
      <button
        className="account-btn"
        onClick={[Function]}
        type="button"
      >
        DELETE ACCOUNT
      </button>
      <button
        className="logout-btn"
        onClick={[Function]}
        type="button"
      >
        LOGOUT
      </button>
    </div>
  </div>
  <div
    className="dashboard"
  >
    <h1
      className="header-left"
    >
      Analytics
    </h1>
    <div
      className="dash-flex"
    >
      <div
        className="dash-card"
      >
        <h4>
          TOTAL COMPENSATION EARNED
        </h4>
        <p>
          $0
        </p>
      </div>
      <div
        className="dash-card"
      >
        <h4>
          TOTAL ENROLLED STUDIES
        </h4>
        <p>
          0 studies
        </p>
      </div>
      <div
        className="dash-card"
      >
        <h4>
          TOTAL RECOMMENDED STUDIES
        </h4>
        <p>
          0 studies
        </p>
      </div>
      <div
        className="dash-card"
      >
        <h4>
          TOTAL MESSAGES SENT
        </h4>
        <p>
          0 messages
        </p>
      </div>
    </div>
  </div>
</div>
`;
